tuta-sdk/rust/src/crypto/entity_facade.rs

```diff
@@ -0,0 +1,300 @@
+                (ElementValue::Null, Cardinality::One) => Err(ApiCallError::InternalSdkError { error_message: format!("Value {association_name} with cardinality ONE can't be null") }),
+                (ElementValue::Array(arr), Cardinality::Any) => {
+                    let mut aggregate_vec: Vec<ElementValue> = Vec::new();
+                    for (index, aggregate) in arr.into_iter().enumerate() {
+                        match aggregate {
+                            ElementValue::Dict(entity) => {
+                                let mut decrypted_aggregate = match self.decrypt_and_map(aggregate_type_model, entity, session_key) {
+                                    Ok(dec_aggregate) => dec_aggregate,
+                                    Err(error) => return Err(error)
+                                };
+
+                                // Errors should be grouped inside the top-most object, so they should be
+                                // extracted and removed from aggregates
+                                if decrypted_aggregate.get("errors").is_some() {
+                                    let error_key = &format!("{}_{}", association_name, index);
+                                    self.extract_errors(error_key, &mut errors, &mut decrypted_aggregate);
+                                }
+
+                                aggregate_vec.push(ElementValue::Dict(decrypted_aggregate));
+                            }
+                            _ => return Err(ApiCallError::InternalSdkError { error_message: format!("Invalid aggregate format. {} isn't a dict", association_name) })
+                        }
+                    };
+
+                    Ok((ElementValue::Array(aggregate_vec), errors))
+                }
+                (ElementValue::Dict(dict), Cardinality::One | Cardinality::ZeroOrOne) => {
+                    let decrypted_aggregate = self.decrypt_and_map(aggregate_type_model, dict, session_key);
+                    match decrypted_aggregate {
+                        Ok(mut dec_aggregate) => {
+                            self.extract_errors(association_name, &mut errors, &mut dec_aggregate);
+                            Ok((ElementValue::Dict(dec_aggregate), errors))
+                        }
+                        Err(_) => {
+                            return Err(ApiCallError::InternalSdkError { error_message: format!("Failed to decrypt association {association_name}") });
+                        }
+                    }
+                }
+                _ => Err(ApiCallError::InternalSdkError { error_message: format!("Invalid association {association_name}") })
+            }
+        } else {
+            Ok((association_data, errors))
+        };
+    }
+
+    fn extract_errors(&self, association_name: &String, errors: &mut HashMap<String, ElementValue>, dec_aggregate: &mut ParsedEntity) {
+        if dec_aggregate.get("errors").is_some() {
```

Comment:
not important: you can probably do both `get` and `remove` with just `remove`  and `match`